### PR TITLE
NSFS | NC | fix list-objecs-versions issues

### LINKED ISSUE #7372: GAP | NSFS | Versioning | List Object Version - Always Returns `IsLatest` true
### Environment info
- NooBaa Version: current master (5.14)
- Platform: Rancher Desktop 1.19.0 (VM and cluster)

### Actual behavior
1. We always return `IsLatest` true.

### Expected behavior
1. Only the latest version should return `IsLatest` true.

#### General Settings:
1. Change the configuration of NSFS Versioning to be Enabled in file `config.js` and change `config.NSFS_VERSIONING_ENABLED` to `true`.
2. Deploy noobaa system on minikube (Based on the instructions [here](https://gist.github.com/dannyzaken/c85f0006e5d6582c8d8666cafce6ab76) - the steps of _‘Build images’_ and _‘Deploy noobaa’_).
_Note: `nb` is an alias that runs the local operator from `build/_output/bin` (alias created by `devenv`)._
4. Deploy NSFS on minikube (Based on the instructions [here](https://github.com/noobaa/noobaa-core/wiki/NSFS-on-Kubernetes/8fa95c3b827c40f2435a63f60fae5f7285e6b221)).
#### Steps:
5. Enabling versioning
`aws --endpoint-url=https://localhost:12443 --no-verify-ssl s3api put-bucket-versioning --bucket fs1-jenia-bucket --versioning-configuration Status=Enabled`
6. Put an object twice
` s3-nb-user-1 aws --endpoint-url=https://localhost:12443 --no-verify-ssl s3api put-object --bucket fs1-jenia-bucket --key fish.txt --body <some_txt_file>`
` s3-nb-user-1 aws --endpoint-url=https://localhost:12443 --no-verify-ssl s3api put-object --bucket fs1-jenia-bucket --key fish.txt --body <some_txt_file>`
7. List object versions:
`aws --endpoint-url=https://localhost:12443 --no-verify-ssl s3api list-object-versions --bucket fs1-jenia-bucket`

### More information - Screenshots / Logs / Other output
Noobaa output:
Screenshot 2023-07-02 at 16 58 36

AWS output:
Screenshot 2023-07-02 at 17 09 36

### PR DESCRIPTION
**Pull request by nadavMiz**
### Explain the changes
1. change _get_object_list to return version id in suspended mode, causing list-object-versions to return version-id in suspended mode as well
2. set isLatest flags for objects in the main bucket folder and not in .version
3. add special case to check if delete marker is the latest version by checking if its the most recent entry of the key

### Testing Instructions:
1. sudo NC_CORETEST=true node ./node_modules/mocha/bin/mocha src/test/unit_tests/test_bucketspace.js

- [ ] Doc added/updated
- [x] Tests added

### CODE REVIEW

File: src/sdk/namespace_fs.js [hunk 1]
```diff
@@ -869,7 +872,7 @@ class NamespaceFS {
                 if (r.common_prefix) {
                     res.common_prefixes.push(r.key);
                 } else {
-                    obj_info = this._get_object_info(bucket, r.key, r.stat, 'null', false, true);
+                    obj_info = this._get_object_info(bucket, r.key, r.stat, 'null', false, r.is_latest);
```
Comment (romayalon) [inline src/sdk/namespace_fs.js:875]: Unrelated to your changes but you change this line anyway, 
I remember I wrote myself to change this 'null' to a boolean, does it make sense?

File: src/sdk/namespace_fs.js [hunk 2]
```diff
@@ -2293,16 +2296,17 @@ class NamespaceFS {
...
+     * @param {string} key
+     * @param {nb.NativeFSStats} stat
      * @returns {nb.ObjectInfo}
      */
     _get_object_info(bucket, key, stat, return_version_id, isDir, is_latest = true) {
         const etag = this._get_etag(stat);
         const create_time = stat.mtime.getTime();
         const encryption = this._get_encryption_info(stat);
-        const version_id = return_version_id && this._is_versioning_enabled() && this._get_version_id_by_xattr(stat);
+        const version_id = return_version_id && (this._is_versioning_enabled() || this._is_versioning_suspended()) &&
```
[the remark this answers is not in the document]
Comment (nadavMiz) [reply src/sdk/namespace_fs.js]: yes, its always true and its value is ignored. I removed this value

File: src/sdk/namespace_fs.js (hunk 1 above)
Comment (nadavMiz) [reply src/sdk/namespace_fs.js]: its the return_version_id value its ignored, so I removed it

File: src/test/unit_tests/test_bucketspace_versioning.js [hunk 3]
```diff
@@ -3101,6 +3102,31 @@ mocha.describe('List-objects', function() {
...
+    mocha.it('list object versions - should show version-id in suspention mode', async function() {
+        await s3_client.putBucketVersioning({ Bucket: bucket_name, VersioningConfiguration: { MFADelete: 'Disabled', Status: 'Suspended' } });
+        const res = await s3_client.listObjectVersions({Bucket: bucket_name});
+        let count = 0;
+        res.Versions.forEach(val => {
+            if (val.VersionId === 'null') {
+                count += 1;
+            }
+        });
+        assert.equal(count, 5);
```
[the remark this answers is not in the document]
Comment (nadavMiz) [reply src/test/unit_tests/test_bucketspace_versioning.js:3128]: there was a mistake, I changed it to `val.VersionId !== 'null'`. there should be 5 entries with real version, and one with null version. I check for that

File: src/test/unit_tests/test_bucketspace_versioning.js [hunk 4]
```diff
@@ -3101,6 +3102,31 @@ mocha.describe('List-objects', function() {
             assert.notEqual(obj.Prefix, ".versions/");
         });
     });
+
+    mocha.it('list object versions - should show isLatest flag only for latest objects', async function() {
+        const res = await s3_client.listObjectVersions({Bucket: bucket_name});
+        res.Versions.forEach(val => {
+            if (val.IsLatest) {
```
[the remark this answers is not in the document]
Comment (nadavMiz) [reply src/test/unit_tests/test_bucketspace_versioning.js:3109]: there are two keys in this test, key and dir_key. I check that the latest version of each of those is the correct one

[the remark this answers is not in the document]
Comment (romayalon) [reply src/test/unit_tests/test_bucketspace_versioning.js:3109]: but if all are false? how would you fail this test?

File: src/sdk/namespace_fs.js [hunk 5]
```diff
@@ -848,6 +851,19 @@ class NamespaceFS {
...
             };
 
+            let previous_key;
+            /**
+             * delete markers are always in the .versions folder, so we need to have special case to determine
+             * if they are delete markers. since the result list is ordered by latest entries first, the first
+             * entry of every key is the latest
+             * @param {Object} obj_info
+             */
+            const set_latest_delete_marker = obj_info => {
```
Comment (romayalon) [inline src/sdk/namespace_fs.js:861]: @nadavMiz We need to figure out how we can get isLatest info if we move to the new structure or in an unordered list...

Comment (romayalon) [reply src/sdk/namespace_fs.js:861]: @naveenpaul1 on your PR do you already support list object versions unordered as well?

Comment (romayalon) [reply src/sdk/namespace_fs.js:861]: @naveenpaul1 maybe in the future

**Review by romayalon — APPROVED** [review]

**Comment by rkomandu** [comment]
is this backported to 4.17 , in noobaa d/s 0904, this one is still a problem ""IsLatest": true"

**Comment by romayalon** [comment]
@rkomandu it was backported on 0904, so no, it's not there...

**Pull request merged**